src/EventStore.Core/TransactionLog/Chunks/TFChunk/TFChunk.cs

```diff
@@ -979,38 +1012,44 @@ public enum ChunkVersions : byte {
 		}
 
 		private bool TryDestructMemStreams() {
-			var writerWorkItem = _writerWorkItem;
-			if (writerWorkItem != null)
-				writerWorkItem.DisposeMemStream();
+			lock (_cachedDataLock) {
+				if (_cacheStatus != CacheStatus.Uncaching && !_selfdestructin54321)
+					return false;
 
-			int memStreamCount = Interlocked.CompareExchange(ref _memStreamCount, 0, 0);
-			var wasZero = memStreamCount == 0;
+				var writerWorkItem = _writerWorkItem;
+				if (writerWorkItem != null)
+					writerWorkItem.DisposeMemStream();
 
-			ReaderWorkItem workItem;
-			while (_memStreams.TryDequeue(out workItem)) {
-				memStreamCount = Interlocked.Decrement(ref _memStreamCount);
-			}
+				int memStreamCount = Interlocked.CompareExchange(ref _memStreamCount, 0, 0);
+
+				ReaderWorkItem workItem;
+				while (_memStreams.TryDequeue(out workItem)) {
+					memStreamCount = Interlocked.Decrement(ref _memStreamCount);
+				}
+
+				if (memStreamCount < 0)
+					throw new Exception("Count of memory streams reduced below zero.");
 
-			if (memStreamCount < 0)
-				throw new Exception("Count of memory streams reduced below zero.");
-			if (memStreamCount == 0)
-			{
-				if (!wasZero) {
-					// we are the last who should "turn the light off" for memory streams
+				if (memStreamCount == 0) {
+					// make sure "the light is off" for memory streams
 					FreeCachedData();
+					return true;
 				}
 
-				return true;
+				return false;
 			}
-
-			return false;
 		}
 
 		private void FreeCachedData() {
-			var cachedData = Interlocked.Exchange(ref _cachedData, IntPtr.Zero);
-			if (cachedData != IntPtr.Zero) {
-				Marshal.FreeHGlobal(cachedData);
-				GC.RemoveMemoryPressure(_cachedLength);
+			lock (_cachedDataLock) {
+				var cachedData = _cachedData;
```

Comment:
can we get rid of the `cachedData` variable, now that it's within a lock?

Reply:
yeah we could